keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1655 @@
+    //
+    // The apiserver does not check that the referenced class
+    // exists, but a driver-specific admission webhook
+    // may require that and is allowed to reject claims where
+    // the class is missing.
+    ResourceClassName string
+
+    // Parameters holds arbitrary values that will be available to the driver
+    // when allocating a resource for the claim.
+    Parameters runtime.RawExtension
+
+    // Allocation can start immediately or when a Pod wants to use
+    // the resource. Waiting for a Pod is the default.
+    AllocationMode AllocationMode
+}
+
+// AllocationMode describes whether a ResourceClaim gets allocated
+// immediately when it gets created (AllocationModeImmediate)
+// or whether allocation is delayed until it is needed for
+// a Pod (AllocationModeDelayed). Other modes might get
+// added in the future.
+type AllocationMode string
+
+const (
+     // When a ResourceClass has AllocationModeImmediate,
+     // allocation of ResourceClaims with that ResourceClass
+     // starts as soon as the ResourceClaim gets created.
+     // This is done without considering the needs of
+     // Pods that will use the ResourceClaim because those
+     // Pods are not known yet.
+     AllocationModeImmediate AllocationMode = “Immediate”
+
+     // When a ResourceClass has AllocationModeDelayed,
+     // allocation of ResourceClaims with that ResourceClass
+     // is delayed until a Pod gets scheduled that needs the
+     // ResourceClaim. The scheduler will consider all
+     // resource requirements of that Pod and trigger
+     // allocation for a node that fits the Pod.
+     AllocationModeDelayed AllocationMode = “Delayed”
+}
+
+type ResourceClaimStatus struct {
+   // Explains what the current status of the claim is and
+   // determines which component needs to do something.
+   Phase ResourceClaimPhase
+
+   // A copy of the driver name from the ResourceClass at
+   // the time when allocation started. Drivers can
+   // filter claims by this field. It's also necessary to
+   // support deallocation when the class gets deleted before
+   // a claim.
+   DriverName string
+
+   // When allocation is delayed, the scheduler must set
+   // the node for which it wants the resource to be allocated
+   // before the driver proceeds with allocation.
+   //
+   // For immediate allocation, the scheduler will not set
+   // this field. The resource driver controller may
+   // set it to trigger allocation on a specific node if the
+   // resources are local to nodes.
+   SelectedNode string
+
+   // When allocation is delayed, and the scheduler needs to
+   // decide on which node a Pod should run, it will first
+   // ask the driver on which nodes the resource might be
+   // made available. To trigger that check, the scheduler
+   // provides the names of nodes which might be suitable
+   // for the Pod. Will be updated periodically until
+   // the claim is allocated.
+   PotentialNodes []string
```

Comment:
is this list unbounded?  what happens with large clusters?

Reply:
This is indeed a problem. Scheduler extender webhooks have the same issue, but here it is perhaps more critical because the object gets handled by the apiserver.

What might work is to select some potential nodes randomly. It reduces the chance of finding a suitable node because the response by the driver cannot rule out a node candidate that it doesn't know about.